src/graphene.cpp

```diff
@@ -564,22 +563,30 @@ bool CGrapheneBlock::process(CNode *pfrom,
         return true;
     }
 
-    // These must be checked outside of the mempool.cs lock or deadlock may occur.
+    // This must be checked outside of the mempool.cs lock or deadlock may occur.
     // A merkle root mismatch here does not cause a ban because and expedited node will forward an graphene
     // without checking the merkle root, therefore we don't want to ban our expedited nodes. Just request
     // a failover block if a mismatch occurs.
-    // Also, there is a remote possiblity of a Tx hash collision therefore if it occurs we request a failover
-    // block.
-    if (collision || !fMerkleRootCorrect)
+    if (!fMerkleRootCorrect)
     {
         RequestFailoverBlock(pfrom, header.GetHash());
         graphenedata.ClearGrapheneBlockData(pfrom, header.GetHash());
 
-        if (!fMerkleRootCorrect)
-            return error(
-                "Mismatched merkle root on grapheneblock: requesting failover block, peer=%s", pfrom->GetLogName());
-        else
-            return error("TX HASH COLLISION for grapheneblock: requesting a full block, peer=%s", pfrom->GetLogName());
+        return error(
+            "Mismatched merkle root on grapheneblock: requesting failover block, peer=%s", pfrom->GetLogName());
+    }
+
+    // This also must be checked outside of the mempool.cs lock or deadlock may occur.
+    // There is a remote possiblity of a Tx hash collision, which would also occur for XThin,
+    // therefore if it occurs we request a full block.
+    if (collision)
```

Comment:
with xthin where this happen we request a thinblock (which use full 64bit hashes) rather than a xthinblock (which use cheap hashes).

Reply:
Yep, @ptschip just alerted me to this. I'm going to rollback this commit and add a comment to the earlier code that we will allow XThin to fail, which will itself failover to a thinblock.